backend/service/auditsink/slack/slack.go

```diff
@@ -104,3 +124,96 @@ func formatText(username string, event *auditv1.RequestEvent) string {
 
 	return messageText
 }
+
+// FormatCustomText parses out the audit event metdata request for the custom slack message text
+func FormatCustomText(message string, event *auditv1.RequestEvent) (string, error) {
+	tmpl, err := template.New("customText").Funcs(funcMap).Parse(message)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := getAuditTemplateData(event)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, data); err != nil {
+		return "", err
+	}
+
+	// When a value is nil, the Go Template returns "<no value>" in its place. Replacing this with null instead.
+	// We hit this scenario with https://github.com/lyft/clutch/blob/main/api/k8s/v1/k8s.proto#L860
+	sanitized := strings.ReplaceAll(buf.String(), "<no value>", "null")
+
+	return sanitized, nil
+}
+
+// returns the API request/response details saved in an audit event
+func getAuditTemplateData(event *auditv1.RequestEvent) (*auditTemplateData, error) {
+	// proto -> json
+	reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+	respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var requestMetadata map[string]interface{}
+	var responseMetadata map[string]interface{}
+	// json -> map[string]interface{}
+	err = json.Unmarshal(reqJSON, &requestMetadata)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(respJSON, &responseMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return &auditTemplateData{
+		Request:  requestMetadata,
+		Response: responseMetadata,
+	}, nil
+}
+
+// for inputs that are type slice/map, returns a formatted slack list OR "None" if the slice/map is empty
+func slackList(data interface{}) string {
+	if data == nil {
```

Comment:
forgot to add a nil check, otherwise will panic later at `reflect.TypeOf(data).Kind()`. added two test cases for this: https://github.com/lyft/clutch/pull/1093/files#diff-17fcfcb6bbd591d6dcd0e04f6a9ceb6bb8b2d0cba8ee7ecf84f1cf907ae73dccR104, https://github.com/lyft/clutch/pull/1093/files#diff-17fcfcb6bbd591d6dcd0e04f6a9ceb6bb8b2d0cba8ee7ecf84f1cf907ae73dccR198